keanu-project/src/test/java/io/improbable/keanu/e2e/regression/RadonHeirarchicalRegression.java

```diff
@@ -0,0 +1,179 @@
+
+        RegressionModel model = RegressionModel.
+            withTrainingData(x, y).
+            withRegularization(RegressionRegularization.RIDGE).
+            withPriorOnWeights(0., 5.).
+            withPriorOnIntercept(0., 5.).
+            build();
+
+        model.observe();
+        model.fit();
+
+        return model;
+    }
+
+    private void buildHeirarchicalNetwork(List<Data> radonData, int numberOfModels) {
+        GaussianVertex muAlpha = new GaussianVertex(new long[]{1, 1}, 0, 100).setLabel("MuIntercept");
+        GaussianVertex muBeta = new GaussianVertex(new long[]{1, 1}, 0, 100).setLabel("MuGradient");
+
+        HalfGaussianVertex sigmaAlpha = new HalfGaussianVertex(new long[]{1, 1}, 10.).setLabel("SigmaIntercept");
+        HalfGaussianVertex sigmaBeta = new HalfGaussianVertex(new long[]{1, 1}, 10.).setLabel("SigmaGradient");
+
+        int numPartitions = radonData.size() / numberOfModels;
+
+        double[] allFloor = radonData.stream().mapToDouble(k -> k.floor).toArray();
+        double[] allRadon = radonData.stream().mapToDouble(k -> k.log_radon).toArray();
+
+        List<RegressionModel> models = new ArrayList<>();
+
+        for (int i = 0; i < numberOfModels; i++) {
+            RegressionModel model = createSubModel(allFloor, allRadon, i, numPartitions, muBeta, muAlpha, sigmaBeta, sigmaAlpha);
+            models.add(model);
+        }
+
+        muAlpha.setValue(1.);
+        sigmaAlpha.setValue(0.5);
+
+        muBeta.setValue(-1.);
+        sigmaBeta.setValue(0.5);
+
+        optimise(new BayesianNetwork(muAlpha.getConnectedGraph()), models);
+    }
+
+    private RegressionModel createSubModel(double[] allFloor,
+                                           double[] allRadon,
+                                           int i,
+                                           int size,
+                                           DoubleVertex muGradient,
+                                           DoubleVertex muIntercept,
+                                           DoubleVertex sigmaGradient,
+                                           DoubleVertex sigmaIntercept) {
+
+        int startIndex = size * i;
+        int endIndex = size * (i + 1);
+
+        double[] floorForSubModel = Arrays.copyOfRange(allFloor, startIndex, endIndex);
+        double[] radonForSubModel = Arrays.copyOfRange(allRadon, startIndex, endIndex);
+
+        DoubleTensor x = DoubleTensor.create(floorForSubModel, 1, floorForSubModel.length);
+        DoubleTensor y = DoubleTensor.create(radonForSubModel, 1, floorForSubModel.length);
+
+        RegressionModel model = RegressionModel.
+            withTrainingData(x, y).
+            withRegularization(RegressionRegularization.RIDGE).
+            withPriorOnWeights(muGradient, sigmaGradient).
+            withPriorOnIntercept(muIntercept, sigmaIntercept).
+            build();
+
+        model.observe();
+
+        return model;
```

Comment:
so what's the deal with these models? They never get fitted?

Reply:
Nope - these are the sub-models that are connected together through the parent `random variable`, we `fit` the parent model